include/dali/core/tensor_layout.h

```diff
@@ -0,0 +1,475 @@
+   * allowing 1 extra character to be stored.
+   */
+  char data_[max_ndim + 1] = { 0 };
+  DALI_HOST_DEV
+  void set_size(int n) {
+    assert(n >= 0 && n <= max_ndim);
+    data_[max_ndim] = max_ndim - n;
+  }
+  DALI_HOST_DEV
+  friend constexpr TensorLayout operator+(const TensorLayout &a, const TensorLayout &b);
+};
+
+static_assert(sizeof(TensorLayout) == 16, "Tensor layout size should be exactly 16B");
+
+DALI_HOST_DEV
+constexpr TensorLayout operator+(const TensorLayout &a, const TensorLayout &b) {
+  assert(a.size() + b.size() < TensorLayout::max_ndim);
+  TensorLayout result = a;
+  int i = 0, j = 0;
+  for (i = result.ndim(), j = 0; i < TensorLayout::max_ndim && j < b.ndim(); i++, j++)
+    result[i] = b[j];
+  result[i] = 0;
+
+  /* Cannot use
+   *   result.set_size(i);
+   * because it's not constexpr
+   */
+  result.data_[TensorLayout::max_ndim] = TensorLayout::max_ndim - i;
+  return result;
+}
+
+
+DALI_HOST_DEV
+constexpr TensorLayout TensorLayout::skip(char dim_name) const noexcept {
+  int i = find(dim_name);
+  if (i < 0)
+    return *this;
+  return first(i) + sub(i+1);
+}
+
+#define DEFINE_TENSOR_LAYOUT_COMPARISON(op)                             \
+inline bool operator op(const TensorLayout &tl, const std::string &s) { \
+  return std::strcmp(tl.c_str(), s.c_str()) op 0;                       \
+}                                                                       \
+DALI_HOST_DEV                                                           \
+constexpr bool operator op(const TensorLayout &tl, const char *s) {     \
+  return tl.compare(s) op 0;                                            \
+}                                                                       \
+inline bool operator op(const std::string &s, const TensorLayout &tl) { \
+  return std::strcmp(s.c_str(), tl.c_str()) op 0;                       \
+}                                                                       \
+DALI_HOST_DEV                                                           \
+constexpr bool operator op(const char *s, const TensorLayout &tl) {     \
+  return TensorLayout(s).compare(tl) op 0;                              \
+}
+
+DEFINE_TENSOR_LAYOUT_COMPARISON(<)  // NOLINT
+DEFINE_TENSOR_LAYOUT_COMPARISON(>)  // NOLINT
+DEFINE_TENSOR_LAYOUT_COMPARISON(<=)
+DEFINE_TENSOR_LAYOUT_COMPARISON(>=)
+DEFINE_TENSOR_LAYOUT_COMPARISON(==)
+DEFINE_TENSOR_LAYOUT_COMPARISON(!=)
+
+/** Provides basic functions for querying TensorLayout properties */
+struct LayoutInfo {
+  DALI_HOST_DEV
+  static bool HasSampleDim(const TensorLayout &tl) {
+    // sample dim may only appear at outermost level
+    return !tl.empty() && tl[0] == 'N';
+  }
+  DALI_HOST_DEV
+  static int DimIndex(const TensorLayout &tl, char dim_symbol) {
+    return tl.find(dim_symbol);
+  }
+};
+
+/** @brief Provides functions for querying TensorLayout properties,
+ *         assuming that the layout describes an image
```

Comment:
This certainly doesn't play well with 16:9 screens :(

Reply:
We have 16:10 at the office ;)

Reply:
It still doesn't work until we rotate them to 10:16.

Reply:
We agreed on one style of docs, I'd kindly request you stick to it. Otherwise, there's no point of having this `lint` stuff, if we don't stick to our own rules...